File: python/Day3/day3.py
import time
target = 361527


direction = 0
places = [(0, 0)]
i = 1
step_length = 1
step = 0
stepped_once = False
values = {(0,0):1}
found = False
part2 = 0
while i < target:

    x, y = places[-1]

    if direction == 0:
        places.append((x + 1, y))
    elif direction == 1:
        places.append((x, y + 1))
    elif direction == 2:
        places.append((x - 1, y))
    else:
        places.append((x, y - 1))

    #Part 2
    new_value = sum([values[(x + a, y + b)] for a in [-1,0,1] for b in [-1,0,1] if (x + a, y + b) in values])
    values[(x,y)] = new_value
    if not found and new_value > 361527:
        part2 = new_value
        found = True

    step += 1
    i += 1
    if step == step_length:
        if stepped_once:
            step_length += 1
            stepped_once = False
        else:
            stepped_once = True
        direction = (direction + 1) % 4
        step = 0

print('Part1:', sum(map(abs, places[-1])))
print('Part2:', part2)

File: kotlin/src/main/AoC2017/day3.kt
package main.AoC2017

import kotlin.math.abs
import kotlin.system.measureTimeMillis


fun main(args: Array<String>) {

    
    val directions = listOf(Pair(1, 0), Pair(0, 1), Pair(-1, 0), Pair(0, -1))
    val points = mutableListOf(Pair(0, 0))
    val values = hashMapOf(Pair(0, 0) to 1)
    val target = 361527
    var direction = 0
    var stepLength = 1

    while (points.size < target) {
        for (times in 0 until 2) {
            for (steps in 0 until stepLength) {
                var (x, y) = points[points.lastIndex]
                var (dx, dy) = directions[direction]
                val currentPoint = Pair(x + dx, y + dy)
                points.add(currentPoint)
                val value = getAllNeighbors(currentPoint)
                        .map { values[it] }
                        .filterNotNull()
                        .sum()
                values[currentPoint] = value
            }
            direction = (direction + 1) % 4
        }
        stepLength++
    }

    println("Part 1: ${points[target - 1].toList().map { abs(it) }.sum()}")
    println("Part 2: ${values.values.filter { it > target }.sorted().first()}")

}

fun getAllNeighbors(point: Pair<Int, Int>): List<Pair<Int, Int>> {
    val neighbors = mutableListOf<Pair<Int, Int>>()
    val (x, y) = point
    for (dx in -1..1)
        for (dy in -1..1)
            if (dx != 0 || dy != 0)
                neighbors.add(Pair(x + dx, y + dy))
    return neighbors
}